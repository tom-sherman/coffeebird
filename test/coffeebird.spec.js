/* global it, expect */
const { transpile } = require('../src/coffeebird')
const { loadExample } = require('./util')

it('should transpile examples', () => {
  const helloWorld = loadExample('hello-world.coffeebird')
  expect(transpile(helloWorld)).toMatchSnapshot()

  const helloWorldExtended = loadExample('hello-world-extended.coffeebird')
  expect(transpile(helloWorldExtended)).toMatchSnapshot()

  const comments = loadExample('comments.coffeebird')
  expect(transpile(comments)).toMatchSnapshot()

  const rules = loadExample('rules.coffeebird')
  expect(transpile(rules)).toMatchSnapshot()
})
