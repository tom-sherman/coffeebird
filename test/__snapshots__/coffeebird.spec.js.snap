// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should transpile examples 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<rbl:kb xmlns:rbl=\\"http://rbl.io/schema/RBLang\\">
	<concept name=\\"Person\\" type=\\"string\\" />
	<concept name=\\"Country\\" type=\\"string\\" />
	<concept name=\\"Language\\" type=\\"string\\" />

	<rel name=\\"speaks\\" subject=\\"Person\\" object=\\"Language\\" plural=\\"false\\" allowUnknown=\\"false\\" askable=\\"all\\" />
	<rel name=\\"lives in\\" subject=\\"Person\\" object=\\"Country\\" plural=\\"false\\" allowUnknown=\\"false\\" askable=\\"all\\" />
	<rel name=\\"national language\\" subject=\\"Country\\" object=\\"Language\\" plural=\\"false\\" allowUnknown=\\"false\\" askable=\\"none\\" />

	<concinst name=\\"Fred\\" type=\\"Person\\" />
	<concinst name=\\"English\\" type=\\"Language\\" />
	<concinst name=\\"French\\" type=\\"Language\\" />
	<concinst name=\\"German\\" type=\\"Language\\" />
	<concinst name=\\"England\\" type=\\"Country\\" />
	<concinst name=\\"France\\" type=\\"Country\\" />
	<concinst name=\\"Germany\\" type=\\"Country\\" />

	<relinst type=\\"national language\\" subject=\\"England\\" object=\\"English\\" cf=\\"100\\" />
	<relinst type=\\"national language\\" subject=\\"France\\" object=\\"French\\" cf=\\"100\\" />
	<relinst type=\\"lives in\\" subject=\\"Fred\\" object=\\"England\\" cf=\\"100\\" />

	<relinst type=\\"speaks\\" cf=\\"75\\" minimumRuleCertainty=\\"60\\">
		<condition rel=\\"lives in\\" subject=\\"%S\\" object=\\"%COUNTRY\\" weight=\\"100\\" behaviour=\\"mandatory\\" />
		<condition rel=\\"national language\\" subject=\\"%COUNTRY\\" object=\\"%O\\" weight=\\"100\\" behaviour=\\"mandatory\\" />
	</relinst>

</rbl:kb>"
`;

exports[`should transpile examples 2`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<rbl:kb xmlns:rbl=\\"http://rbl.io/schema/RBLang\\">
<!--  The purpose of this example is to show all of the features of Coffeebird. -->

	<concept name=\\"Person\\" type=\\"string\\" />
	<concept name=\\"Country\\" type=\\"string\\" />
	<concept name=\\"Language\\" type=\\"string\\" />

	<rel name=\\"speaks\\" subject=\\"Person\\" object=\\"Language\\" plural=\\"false\\" allowUnknown=\\"false\\" askable=\\"all\\" />
	<rel name=\\"lives in\\" subject=\\"Person\\" object=\\"Country\\" plural=\\"false\\" allowUnknown=\\"false\\" askable=\\"all\\" />
	<rel name=\\"national language\\" subject=\\"Country\\" object=\\"Language\\" plural=\\"false\\" allowUnknown=\\"false\\" askable=\\"secondFormObject\\" />

	<concinst name=\\"Fred\\" type=\\"Person\\" />
	<concinst name=\\"English\\" type=\\"Language\\" />
	<concinst name=\\"French\\" type=\\"Language\\" />
	<concinst name=\\"German\\" type=\\"Language\\" />
	<concinst name=\\"England\\" type=\\"Country\\" />
	<concinst name=\\"France\\" type=\\"Country\\" />
	<concinst name=\\"Germany\\" type=\\"Country\\" />

	<relinst type=\\"national language\\" subject=\\"England\\" object=\\"English\\" cf=\\"100\\" />
	<relinst type=\\"national language\\" subject=\\"France\\" object=\\"French\\" cf=\\"100\\" />
	<relinst type=\\"lives in\\" subject=\\"Fred\\" object=\\"England\\" cf=\\"100\\" />

	<relinst type=\\"speaks\\" cf=\\"50\\">
		<condition rel=\\"lives in\\" subject=\\"%S\\" object=\\"%COUNTRY\\" weight=\\"0\\" behaviour=\\"mandatory\\" />
		<condition rel=\\"has national language\\" subject=\\"%COUNTRY\\" object=\\"%O\\" weight=\\"100\\" behaviour=\\"mandatory\\" />
	</relinst>

	<relinst type=\\"speaks\\" cf=\\"100\\">
		<condition rel=\\"lives in\\" subject=\\"%S\\" object=\\"%COUNTRY\\" weight=\\"100\\" behaviour=\\"mandatory\\" />
		<condition expression=\\"(%COUNTRY = 'Ireland')\\" weight=\\"100\\" behaviour=\\"mandatory\\" />
		<condition expression=\\"'Irish'\\" value=\\"%O\\" weight=\\"100\\" behaviour=\\"mandatory\\" />
	</relinst>

</rbl:kb>"
`;

exports[`should transpile examples 3`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<rbl:kb xmlns:rbl=\\"http://rbl.io/schema/RBLang\\">
<!--  Comments have to sit on their own line -->
	<concept name=\\"Person\\" type=\\"string\\" />

<!--  Block comments are not supported. -->
<!--  But you can always add multiple comments on separate lines. -->

</rbl:kb>"
`;

exports[`should transpile examples 4`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<rbl:kb xmlns:rbl=\\"http://rbl.io/schema/RBLang\\">
<!--  Let's setup out concepts, relationships, and facts -->

	<concept name=\\"Person\\" type=\\"string\\" />
	<concept name=\\"Country\\" type=\\"string\\" />
	<concept name=\\"Language\\" type=\\"string\\" />
	<concept name=\\"DOB\\" type=\\"date\\" />

	<rel name=\\"speaks\\" subject=\\"Person\\" object=\\"Language\\" plural=\\"false\\" allowUnknown=\\"false\\" askable=\\"all\\" />
	<rel name=\\"lives in\\" subject=\\"Person\\" object=\\"Country\\" plural=\\"false\\" allowUnknown=\\"false\\" askable=\\"all\\" />
	<rel name=\\"national language\\" subject=\\"Country\\" object=\\"Language\\" plural=\\"false\\" allowUnknown=\\"false\\" askable=\\"all\\" />
	<rel name=\\"date of birth\\" subject=\\"Person\\" object=\\"DOB\\" plural=\\"false\\" allowUnknown=\\"false\\" askable=\\"all\\" />

	<concinst name=\\"Fred\\" type=\\"Person\\" />
	<concinst name=\\"English\\" type=\\"Language\\" />
	<concinst name=\\"French\\" type=\\"Language\\" />
	<concinst name=\\"German\\" type=\\"Language\\" />
	<concinst name=\\"England\\" type=\\"Country\\" />
	<concinst name=\\"France\\" type=\\"Country\\" />
	<concinst name=\\"Germany\\" type=\\"Country\\" />
	<concinst name=\\"Shakespearean\\" type=\\"Country\\" />

	<relinst type=\\"national language\\" subject=\\"England\\" object=\\"English\\" cf=\\"100\\" />
	<relinst type=\\"national language\\" subject=\\"France\\" object=\\"French\\" cf=\\"100\\" />
	<relinst type=\\"lives in\\" subject=\\"Fred\\" object=\\"England\\" cf=\\"100\\" />

<!--  Now on to the rules -->

<!--  We've all seen this before -->
<!--  Semicolon on the last condition is optional and is omitted here. -->
	<relinst type=\\"speaks\\" cf=\\"100\\">
		<condition rel=\\"lives in\\" subject=\\"%S\\" object=\\"%COUNTRY\\" weight=\\"100\\" behaviour=\\"mandatory\\" />
		<condition rel=\\"national language\\" subject=\\"%COUNTRY\\" object=\\"%O\\" weight=\\"100\\" behaviour=\\"mandatory\\" />
	</relinst>

<!--  An object specific rule -->
	<relinst type=\\"speaks\\" object=\\"German\\" cf=\\"100\\">
		<condition rel=\\"lives in\\" subject=\\"%S\\" object=\\"'Germany'\\" weight=\\"100\\" behaviour=\\"mandatory\\" />
	</relinst>

<!--  A subject specific rule -->
	<relinst type=\\"speaks\\" subject=\\"Fred\\" cf=\\"100\\">
		<condition rel=\\"lives in\\" subject=\\"'Fred'\\" object=\\"'German'\\" weight=\\"100\\" behaviour=\\"mandatory\\" />
	</relinst>

<!--  A subject and object specific rule -->
	<relinst type=\\"speaks\\" subject=\\"Fred\\" object=\\"French\\" cf=\\"100\\">
		<condition rel=\\"lives in\\" subject=\\"'Fred'\\" object=\\"'Germany'\\" weight=\\"100\\" behaviour=\\"mandatory\\" />
	</relinst>

<!--  Using a transformative expression -->
	<relinst type=\\"speaks\\" cf=\\"100\\">
		<condition expression=\\"'A language'\\" value=\\"%O\\" weight=\\"100\\" behaviour=\\"mandatory\\" />
	</relinst>

<!--  Using a comparative and transformative expression -->
	<relinst type=\\"speaks\\" cf=\\"100\\">
		<condition rel=\\"lives in\\" subject=\\"%S\\" object=\\"%COUNTRY\\" weight=\\"100\\" behaviour=\\"mandatory\\" />
		<condition expression=\\"(%COUNTRY = 'Italy')\\" weight=\\"100\\" behaviour=\\"mandatory\\" />
		<condition expression=\\"'A language'\\" value=\\"%O\\" weight=\\"100\\" behaviour=\\"mandatory\\" />
	</relinst>

<!--  Date functions -->
	<relinst type=\\"speaks\\" object=\\"Shakespearean\\" cf=\\"100\\">
		<condition rel=\\"date of birth\\" subject=\\"%S\\" object=\\"%DOB\\" weight=\\"100\\" behaviour=\\"mandatory\\" />
		<condition expression=\\"today()\\" value=\\"%TODAY\\" weight=\\"100\\" behaviour=\\"mandatory\\" />
		<condition expression=\\"(yearsBetween(%TODAY, %DOB) gt 500)\\" weight=\\"100\\" behaviour=\\"mandatory\\" />
	</relinst>

</rbl:kb>"
`;
